mlfinlab/features/fracdiff.py

```diff
@@ -0,0 +1,169 @@
+    that approaches zero asymptotically.  The side effect of this function is that
+    it leads to negative drift "caused by an expanding window's added weights"
+    (see page 83 AFML)
+
+    When diff_amt is real (non-integer) positive number then it preserves memory.
+
+    The book does not discuss what should be expected if d is a negative real
+    number.  Conceptually (from set theory) negative d leads to set of negative
+    number of elements.  And that translates into a set whose elements can be
+    selected more than once or as many times as one chooses (multisets with
+    unbounded multiplicity) - see http://faculty.uml.edu/jpropp/msri-up12.pdf.
+
+    :param diff_amt: (float) differencing amount
+    :param size: (int) length of the series
+    :return: (ndarray) weight vector
+    """
+
+    weights = [1.]
+    for k in range(1, size):
+        weights_ = -weights[-1] / k * (diff_amt - k + 1)
+        weights.append(weights_)
+
+    weights = np.array(weights[::-1]).reshape(-1, 1)
+    return weights
+
+
+def frac_diff(series, diff_amt, thresh=0.01):
+    """
+    Source: Chapter 5, AFML (section 5.5);
+
+    References:
+    https://www.wiley.com/en-us/Advances+in+Financial+Machine+Learning-p-9781119482086
+    https://wwwf.imperial.ac.uk/~ejm/M3S8/Problems/hosking81.pdf
+    https://en.wikipedia.org/wiki/Fractional_calculus
+
+    The steps are as follows:
+    - Compute weights (this is a one-time exercise)
+    - Iteratively apply the weights to the price series and generate output points
+
+    This is the expanding window variant of the fracDiff algorithm
+    Note 1: For thresh-1, nothing is skipped
+    Note 2: diff_amt can be any positive fractional, not necessarility bounded [0, 1]
+
+    :param series: (pd.Series) a time series that needs to be differenced
+    :param diff_amt: (float) Differencing amount
+    :param thresh: (float) threshold or epsilon
+    :return: (pd.DataFrame) data frame of differenced series
+    """
+
+    # 1. Compute weights for the longest series
+    weights = get_weights(diff_amt, series.shape[0])
+
+    # 2. Determine initial calculations to be skipped based on weight-loss threshold
+    weights_ = np.cumsum(abs(weights))
+    weights_ /= weights_[-1]
+    skip = weights_[weights_ > thresh].shape[0]
+
+    # 3. Apply weights to values
+    output_df = {}
+    for name in series.columns:
+        series_f = series[[name]].fillna(method='ffill').dropna()
+        output_df_ = pd.Series(index=series.index)
+
+        for iloc in range(skip, series_f.shape[0]):
+            loc = series_f.index[iloc]
+
+            # At this point all entries are non-NAs so no need for the following check
+            # if np.isfinite(series.loc[loc, name]):
+            output_df_[loc] = np.dot(weights[-(iloc + 1):, :].T, series_f.loc[:loc])[0, 0]
+
+        output_df[name] = output_df_.copy(deep=True)
+    output_df = pd.concat(output_df, axis=1)
```

Comment:
What is this line doing? Transposing? We are concatenating a single series.

Reply:
Suppose the input data belonged to several stocks and we wanted to compute fractionally differenced series for all of them.  To meet that need, output_df was created as a dict with the name of the stock as the key.  The loop computes the frac-diff for each stock and adds to the dictionary.   This line converts that dictionary into a data frame